src/hotspot/share/gc/shenandoah/shenandoahScanRemembered.hpp

```diff
@@ -0,0 +1,945 @@
+    uint32_t last_card_in_cluster = ((card_at_end < last_card_of_this_cluster) ? card_at_end: last_card_of_this_cluster);
+
+    uint_16 crossing_map_within_cluster = address - cluster_at_start;
+
+    for (uint32_t card_to_update = card_at_start + 1; card_to_update < last_card_in_cluster; card_to_update++)
+      set_crossing_object_start(card_to_update, crossing_map_within_cluster);
+
+    // If the last card region of this cluster is completely spanned
+    // by this new object,  set its crossing object start as well.
+    if (cluster_at_end > cluster_at_start) {
+      set_crossing_object_start(card_to_update++, crossing_map_within_cluster);
+
+      // Now, we have to update all spanned cards that reside within the
+      // following clusters.
+      while (card_to_update < card_at_end)
+        set_crossing_object_start(card_to_update++, CrossingObjectOverflow);
+
+      // Cases:
+      //
+      // 1. The region for card_at_end is not spanned at all because the
+      //    end of the new object aligns with the start of the last
+      //    card region.  In this case, we do nothing with the
+      //    card_at_end entry of the object_starts array.
+      //
+      // 2. The region for card_at_end is spanned partially.  In this
+      //    case, we do nothing with the card_at_end entry of the
+      //    object_starts array.
+      //
+      // 3. Do not need to consider the case that the region for
+      //    card_at_end is spanned entirely.  If the last region
+      //    spanned by a newly registered object is spanned in its
+      //    entirety, then card_at_end will identify the card region
+      //    that follows the object rather than the last region spanned
+      //    by the object.
+
+      // Bottom line: no further work to be done in any of these cases.
+    }
+    // Otherwise, the last card region of this cluster is not
+    // completely spanned by this new object.  Leave the object_starts
+    // array entry alone.  Two cases:
+    //
+    //  a) This newly registered object represents a consolidation of
+    //     multiple smaller objects, not including the object that
+    //     follows the consolidation.
+    //  b) This newly registered object is created by splitting a
+    //     previously existing object.  In this case, the other
+    //     objects resulting from the split will be registered
+    //     separately before it is necessary to lookup any information
+    //     within the object_starts array.
+    //
+
+#endif  // SUPPORT_FOR_GET_CROSSING_OBJECT_START_NO_LONGER_REQUIRED
+
+#else   // IMPLEMENT_THIS_OPTIMIZATION_LATER
+
+    // Do nothing for now as we have a brute-force implementation
+    // of findSpanningObject().
+
+#endif  // IMPLEMENT_THIS_OPTIMIZATION_LATER
+  }
+
+  // What card number corresponds to old-gen heap addresss p.  (If p
+  // does not refer to old-gen memory, the returned value is undefined.)
+  uint32_t cardNoForAddr(HeapWord *p);
```

Comment:
Should be named card_no_for_addr(HeapWord* p) to match style of most other Hotspot code.

Reply:
Removed.